### ISSUE TITLE
Correct client disconnection, in a cycle of automatic reconnection attempts

### ISSUE THREAD

**Issue by Maxbudkov** [shamblett/mqtt_client#231]
Hi, first of all I want to thank for a great package that helped me a lot. Really thanks for that.

Regarding my question:
Please tell me how to properly disconnect the client during the automatic reconnection process? The documentation says that I have to call the disconnect method on the client, but I can't do it.
I use the onDisconnected and onAutoReconnect methods to write connection state information to streams for the bloc pattern (I'm saying this because I didn't understand if these two callbacks can be used simultaneously).
When I am connected to the broker, and when I abnormally disconnect from it (when the Wi-Fi is disconnected), I get the onAutoReconnect call, which is correct. After that I want to finish connection attempts and by clicking on the button I call the client.disconnect () function, it works (as indicated by the corresponding log:
```
I/flutter (24000): 2020-10-28 15: 20: 35.334358 - MqttConnectionHandlerBase :: disconnect - entered
I/flutter (24000): 2020-10-28 15: 20: 35.336016 - MqttConnectionHandlerBase :: _ performConnectionDisconnect entered
I/flutter (24000): EXAMPLE :: OnDisconnected client callback - Client disconnection
I/flutter (24000): EXAMPLE :: OnDisconnected callback is solicited, this is correct
```
).
However, after doing this, connection attempts continue even if I zero out the client. Accordingly, I am getting the error:

```
E/flutter (25014): [ERROR: flutter / lib / ui / ui_dart_state.cc (177)] Unhandled Exception: Bad state: Cannot add new events after calling close
E/flutter (25014): # 0 _BroadcastStreamController.add (dart: async / broadcast_stream_controller.dart: 249: 24)
E/flutter (25014): # 1 EventBus.fire (package: event_bus / event_bus.dart: 61: 22)
E/flutter (25014): # 2 MqttConnectionHandlerBase.autoReconnect (package: mqtt_client / src / connectionhandling / mqtt_client_mqtt_connection_handler_base.dart: 154: 22)
E/flutter (25014): <asynchronous suspension>
E/flutter (25014): # 3 _rootRunUnary (dart: async / zone.dart: 1198: 47)
E/flutter (25014): # 4 _CustomZone.runUnary (dart: async / zone.dart: 1100: 19)
E/flutter (25014): # 5 _CustomZone.runUnaryGuarded (dart: async / zone.dart: 1005: 7)
E/flutter (25014): # 6 CastStreamSubscription._onData (dart: _internal / async_cast.dart: 85: 11)
E/flutter (25014): # 7 _rootRunUnary (dart: async / zone.dart: 1198: 47)
E/flutter (25014): # 8 _CustomZone.runUnary (dart: async / zone.dart: 1100: 19)
E/flutter (25014): # 9 _CustomZone.runUnaryGuarded (dart: async / zone.dart: 1005: 7)
E/flutter (25014): # 10 _BufferingStreamSubscription._sendData (dart: async / stream_impl.dart: 357: 11)
E/flutter (25014): # 11 _BufferingStreamSubscription._add (dart: async / stream_impl.dart: 285: 7)
E/flutter (25014): # 12 _ForwardingStreamSubscription._add (dart: async / stream_pipe.dart: 127: 11)
E/flutter (25014): # 13 _WhereStream._handleData (dart: async / stream_pipe.dart: 199: 12)
E/flutter (25014): # 14 _ForwardingStreamSubscription._handleData (dart: async / stream_pipe.dart: 157: 13)
E/flutter (25014): # 15 _rootRunUnary (dart: async / zone.dart: 1198: 47)
E/flutter (25014): # 16 _CustomZone.runUnary (dart: async / zone.dart: 1100: 19)
E/flutter (25014): # 17 _CustomZone.runUnaryGuarded (dart: async / zone.dart: 1005: 7)
E/flutter (25014): # 18 _BufferingStreamSubscription._sendData (dart: async / stream_impl.dart: 357: 11)
E/flutter (25014): # 19 _DelayedData.perform (dart: async / stream_impl.dart: 611: 14)
E/flutter (25014): # 20 _StreamImplEvents.handleNext (dart: async / stream_impl.dart: 730: 11)
E/flutter (25014): # 21 _PendingEvents.schedule. <Anonymous closure> (dart: async / stream_impl.dart: 687: 7)
E/flutter (25014): # 22 _rootRun (dart: async / zone.dart: 1182: 47)
E/flutter (25014): # 23 _CustomZone.run (dart: async / zone.dart: 1093: 19)
E/flutter (25014): # 24 _CustomZone.runGuarded (dart: async / zone.dart: 997: 7)
E/flutter (25014): # 25 _CustomZone.bindCallbackGuarded. <Anonymous closure> (dart: async / zone.dart: 1037: 23)
E/flutter (25014): # 26 _rootRun (dart: async / zone.dart: 1190: 13)
E/flutter (25014): # 27 _CustomZone.run (dart: async / zone.dart: 1093: 19)
E/flutter (25014): # 28 _CustomZone.runGuarded (dart: async / zone.dart: 997: 7)
E/ flutter (25014): # 29 _CustomZone.bindCallbackGuarded. <Anonymous closure> (dart: async / zone.dart: 1037: 23)
E/ flutter (25014): # 30 _microtaskLoop (dart: async / schedule_microtask.dart: 41: 21)
E/flutter (25014): # 31 _startMicrotaskLoop (dart: async / schedule_microtask.dart: 50: 5)
E/flutter (25014):
```

I don't understand why, after calling the corresponding function that is mentioned in the documentation, it does not work as expected. Perhaps I am doing something wrong? Can you suggest a solution? Thanks a lot for your work.

**Comment by shamblett** [comment]
Auto reconnect is designed to keep trying when triggered. Why are you setting auto reconnect then calling disconnect, if you want to manually disconnect on your wifi event then don't set auto reconnect and just manually disconnect as you are doing now, or allow auto reconnect to re connect, then call disconnect.

**Comment by Maxbudkov** [comment]
Thanks for your reply. I use auto-reconnect as a feature that my application client can enable and cannot turn it off periodically. Okay, let's say I can assume different logic for my application. Then I don't understand the logic behind the automatic reconnection of the reconnect. Are the settings saved and should I re-create the client after a reconnection situation? Because if the auto reconnect happened, if I lose the network signal again, I get an error
`E/flutter (29661): [ERROR: flutter / lib / ui / ui_dart_state.cc (177)] Unhandled Exception: SocketException: OS Error: Software caused connection abort, errno = 103, address = 192.168.4.1, port = 45658`

and nothing else happens. There are no more reconnection attempts. At the same time, once every 1 minute I get this pingRequest:
```
I/flutter (29661): 2020-10-28 16: 26: 39.285273 - MqttConnectionHandlerBase :: sendMessage - MQTTMessage of type MqttMessageType.pingRequest
I/flutter (29661): Header: MessageType = MqttMessageType.pingRequest, Duplicate = false, Retain = false, Qos = MqttQos.atMostOnce, Size = 0
```

**Comment by shamblett** [comment]
What should happen is if you select auto reconnect the client will try and reconnect to the broker with exactly the same parameters as it was originally connected with, you don't need to do anything. If you get a  network error of some kind during auto reconnect the behavior becomes undefined, the client can't cope with all possible network failure cases. 

If auto reconnect is not working for any reason the best thing to do is to not set it, and as you have seen above monitor your ping request responses using the supplied client  callback, if these stop you can manually disconnect and re-connect again yourself, this gives you better control over the process.

From your use case it seems that auto reconnect is not the best option for you and you need more control over the process if your network reliability is as you describe it.

**Comment by Maxbudkov** [comment]
I understand, but I'm talking about something else. If I have a connection loss, the client makes endless reconnection attempts until he succeeds (regardless of the connection state). If the situation repeats (the second time there was a connection loss), then I get a general connection error and nothing else (there are no attempts to reconnect, although the client settings are the same, and the automatic reconnection option is enabled). Actually, this is what I want to ask. Do I understand correctly that automatic reconnection only works once and does not repeat again unless you manually reconnect the client?

**Comment by Maxbudkov** [comment]
Once again, I repeat the procedure and what I expect from the client:

1. First connection to the broker.
2. The connection is established, everything is fine, messages are being exchanged.
3. The connection is lost.
4. The client tries to connect endlessly until it finds the broker again.
5. The connection was restored, the broker was found, the connection was successful, p.2.
6. In case of repeated loss of connection, I receive an error message from the client and there are no more attempts to reconnect.

(Expected outcome) 6. p. 3 - p. 5

Thanks again for wasting your time

**Comment by shamblett** [comment]
OK, thanks for the scenario. Steps 1 to 5 are what should happen.

If step 6 occurs after a complete restoration of connection that's a bug, auto reconnect should always work if invoked again and it completed successfully, could you supply a log for this.

If step 6 occurs whilst an auto reconnect sequence is in progress then yes, this may not work as described above.

I'm not a flutter user but from other issues raised I believe there is a notion of 'background' in flutter where when an app goes into background it loses its network connection, auto reconnect was not really intended for this case, it  is  really for a connected client that suddenly loses its network. If you know you are going into background you should disconnect then re connect when you come out, this is not an unsolicited event, its a known event. So it depends what you mean by 'the connection is lost', is this just network failure or do you know this is going to happen?.

**Comment by Maxbudkov** [comment]
Thanks again for your answers.
Step 6 assumes that I am out of range of the Wi-Fi router and lost my connection (did not go into the background). After I come back, I automatically reconnect to the broker, as it should be. But if I go out of range of the router again and come back, then the reconnection will not happen. The entire log that the IDE gives me in this case is already written by me:

`E/flutter (29661): [ERROR: flutter/lib/ui/ui_dart_state.cc (177)] Unhandled Exception: SocketException: OS Error: Software caused connection abort, errno = 103, address = 192.168.4.1, port = 45658`

Can you explain to me why this is happening? I really do not understand why the auto reconnect flag is active, the connection is lost, but it is the second time and then the reconnection does not occur.

**Comment by shamblett** [comment]
OK thanks, this sounds like a bug I'll see if I can reproduce it.

**Comment by Maxbudkov** [comment]
Thank you very much, I hope we understand each other. If you have any questions, I am always ready to help, I really look forward to solving the problem.

**Comment by easydo247** [comment]
Hi! Any news on this fix?

**Comment by shamblett** [comment]
Not at the moment, I've been looking at this off and on(more off than on in fact) however xmas break is fast approaching which will give me more time to look at this so hopefully we may have more progress before or around xmas.

**Comment by shamblett** [comment]
Just an update, I've got a unit test that repeatedly forces the client to auto reconnect by calling the doAutoReconnect method, this works every time, so I moved onto to forcing the auto reconnect by repeatedly stopping/starting the broker, first time round this works OK, second time round the client doesn't seem to pick up the broker disconnect, i.e. it doesn't go into auto reconnect, it seems to think the disconnect hasn't happened.

OK, this may be a manifestation of the fault, it may just behave like this when using a localhost broker rather than a true remote one which gives you this error as you have provided above 
```
E/flutter (29661): [ERROR: flutter/lib/ui/ui_dart_state.cc (177)] Unhandled Exception: SocketException: OS Error: Software caused connection abort, errno = 103, address = 192.168.4.1, port = 45658
```
It may be there is more than one problem, for now I'll look at why the client doesn't pick up second and subsequent broker stop/starts and see where we go.
